File: FinalProject/app/src/main/java/com/example/finalproject/MainActivity.java
package com.example.finalproject;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.FragmentTransaction;

import android.content.SharedPreferences;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.graphics.Point;
import android.os.Bundle;
import android.os.PersistableBundle;
import android.util.Log;
import android.util.Pair;
import android.view.Display;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;

public class MainActivity extends AppCompatActivity {

    public static Player player;
    public static HashMap<Integer, Integer> chances_of_fight =new HashMap<>();
    public static MapTitle[][] map=new MapTitle[32][32];
    public static Bitmap[] menu=new Bitmap[4];
    public static HashMap<Integer, HashMap<Integer, Enemy>> chances_of_enemy=new HashMap<>();
    public static ArrayList<Enemy> enemies=new ArrayList<>();
    public static HashMap<Integer, ArrayList<Pair<Item, Integer>>> drop=new HashMap<>();
    public static ArrayList<Element> elements=new ArrayList<>();
    public static ArrayList<ManaChannel> mana_channels =new ArrayList<>();
    public static ArrayList<Type> types =new ArrayList<>();
    public static ArrayList<Form> forms =new ArrayList<>();
    public static ArrayList<ManaReservoir> mana_reservoirs =new ArrayList<>();
    public static ArrayList<Research> researches=new ArrayList();
    private static boolean created=false;
    private static Display display;
    private static Resources res;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        display = getWindowManager().getDefaultDisplay();
        res=getResources();
        FragmentTransaction fragmentTransaction= getSupportFragmentManager().beginTransaction();
        fragmentTransaction.add(R.id.map, new Map());
        fragmentTransaction.add(R.id.status, new StatusBar());
        fragmentTransaction.add(R.id.menu, new Menu());
        fragmentTransaction.commit();
        if (savedInstanceState!=null)
        {
            player=new Player(savedInstanceState.getParcelable("player"));
            Log.d("HHH", player.toString());
        }
        else player=new Player(2, 2);
        setInitialData();
    }


    @Override
    protected void onSaveInstanceState(@NonNull @NotNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable("player", player);
        Log.d("LLL", player.toString());
    }

    @Override
    protected void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        player=new Player(savedInstanceState.getParcelable("player"));
        Log.d("LLLL", player.toString());
    }

    protected static void setInitialData(){
        if (created)
            player=new Player(2, 2);
        created=true;
        Point size = new Point();
        display.getSize(size);
        int width = size.x;
        int n=10, m=10;
        Bitmap a=Bitmap.createScaledBitmap(BitmapFactory.decodeResource(res, R.drawable.xxx), width/5*n, width/5*m, false);
        Bitmap[][] b=new Bitmap[n][m];
        for (int i=0;i<n;i++){
            for (int j=0;j<m;j++){
                b[i][j]=Bitmap.createBitmap(a, i*width/5, j*width/5, width/5, width/5);
            }
        }
        for (int i=0;i<map.length;i++){
            for (int j=0;j<map[0].length;j++){
                map[i][j]=new MapTitle(new Pair<>(i, j), Bitmap.createBitmap(width/5, width/5, Bitmap.Config.ARGB_8888), 0);
                if ((i<=1||j<=1)||(i>=map.length-2||j>=map.length-2)){
                    map[i][j].setType(0);
                }
                else{
                    map[i][j].setType(2);
                }
            }
        }
        for (int i=2;i<11;i++){
            for (int j=2;j<11;j++){
                map[j][i].setType(1);
            }
        }
        for (int i=15;i<20;i++){
            for (int j=17;j<22;j++){
                map[i][j].setType(1);
            }
        }
        for (int i=20;i<23;i++){
            for (int j=6;j<9;j++){
                map[i][j].setType(1);
            }
        }
        for (int i=7;i<20;i++){
            map[6][i].setType(5);
        }
        for (int i=20;i<30;i++){
            map[6][i].setType(7);
        }
        for (int i=7;i<21;i++){
            map[i][19].setType(6);
        }
        for (int i=19;i>6;i-=2){
            map[21][i].setType(5);
        }
        map[6][6].setType(3);
        map[17][19].setType(3);
        map[21][7].setType(3);
        for (int i=0;i<map.length;i++){
            for (int j=0;j<map[0].length;j++){
                switch (map[i][j].getType()){
                    case 0: map[i][j].setTexture(b[0][0]); break;
                    case 1: map[i][j].setTexture(b[1][0]); break;
                    case 2: map[i][j].setTexture(b[2][0]); break;
                    case 3: map[i][j].setTexture(b[3][0]); break;
                    case 4: map[i][j].setTexture(b[4][0]); break;
                    case 5: map[i][j].setTexture(b[5][0]); break;
                    case 6: map[i][j].setTexture(b[6][0]); break;
                    case 7: map[i][j].setTexture(b[7][0]); break;
                }
            }
        }
        ArrayList<Research> rqr=new ArrayList<>();
        researches.add(new Research(null, "Basic spell creation", 1, 0, 0, false, true));
        rqr.add(researches.get(0));
        researches.add(new Research((ArrayList<Research>)rqr.clone(), "Fire mage", 3, 1, 1, false, false));
        researches.add(new Research((ArrayList<Research>)rqr.clone(), "Water mage", 3, 1, 2, false, false));
        researches.add(new Research((ArrayList<Research>)rqr.clone(), "Earth mage", 3, 1, 3, false, false));
        researches.add(new Research((ArrayList<Research>)rqr.clone(), "Air mage", 3, 1, 4, false, false));
        drop.put(0, new ArrayList<>());
        drop.get(0).add(new Pair<>(new Item(2, "Rabbit's fur"), 70));
        drop.get(0).add(new Pair<>(new Item(3, "Rabbit's leg"), 30));
        drop.put(1, new ArrayList<>());
        drop.get(1).add(new Pair<>(new Item(5, "Dog's fur"), 100));
        drop.put(2, new ArrayList<>());
        drop.put(3, new ArrayList<>());
        drop.put(4, new ArrayList<>());
        drop.put(5, new ArrayList<>());
        player.setTitle_texture(Bitmap.createBitmap(map[player.getCoordinates().first][player.getCoordinates().second].texture));
        map[player.getCoordinates().first][player.getCoordinates().second].getTexture().eraseColor(Color.BLUE);
        enemies.add(new Enemy("Rabbit", 5, 0, 1, 0, 1, 1, drop.get(0)));
        enemies.add(new Enemy("Dog", 15, 0, 2, 0, 5, 5, drop.get(1)));
        enemies.add(new Enemy("Wolf", 35, 0, 5, 0, 10, 10, drop.get(2)));
        enemies.add(new Enemy("Fox", 20, 0, 3, 0, 7, 6, drop.get(3)));
        enemies.add(new Enemy("Bear", 100, 10, 15, 0, 50, 200, drop.get(4)));
        enemies.add(new Enemy("Highwayman", 50, 10, 20, 0, 100, 50, drop.get(5)));
        chances_of_fight.put(0, 0);
        chances_of_fight.put(1, 20);
        chances_of_fight.put(2, 60);
        chances_of_fight.put(3, 0);
        chances_of_fight.put(4, 10);
        chances_of_fight.put(5, 10);
        chances_of_fight.put(6, 5);
        chances_of_fight.put(7, 5);
        chances_of_enemy.put(1, new HashMap<>());
        chances_of_enemy.get(1).put(30, enemies.get(0));
        chances_of_enemy.get(1).put(70, enemies.get(1));
        chances_of_enemy.put(2, new HashMap<>());
        chances_of_enemy.get(2).put(60, enemies.get(2));
        chances_of_enemy.get(2).put(35, enemies.get(3));
        chances_of_enemy.get(2).put(5, enemies.get(4));
        chances_of_enemy.put(4, new HashMap<>());
        chances_of_enemy.get(4).put(100, enemies.get(5));
        chances_of_enemy.put(5, new HashMap<>());
        chances_of_enemy.get(5).put(100, enemies.get(5));
        chances_of_enemy.put(6, new HashMap<>());
        chances_of_enemy.get(6).put(100, enemies.get(5));
        chances_of_enemy.put(7, new HashMap<>());
        chances_of_enemy.get(7).put(100, enemies.get(5));
        chances_of_enemy.put(8, new HashMap<>());
        chances_of_enemy.get(8).put(100, enemies.get(5));
        elements.add(new Element("Pure mana", -1, 2, false));
        elements.add(new Element("Fire", 0, 10, false));
        elements.add(new Element("Water", 1, 5, false));
        elements.add(new Element("Air", 2, 5, false));
        elements.add(new Element("Earth", 3, 8, false));
        elements.add(new Element("Life", 4, -5, false));
        elements.add(new Element("Death", 5, 5,false));
        elements.add(new Element("Light", 6, 2,false));
        elements.add(new Element("Darkness", 7, 7,false));
        types.add(new Type("Self", 0, true));
        forms.add(new Form("Sphere", 0, true));
        mana_reservoirs.add(new ManaReservoir("Basic", 1, true));
        mana_channels.add(new ManaChannel("Basic", 2, true));
        menu[0]=Bitmap.createScaledBitmap(b[9][0], width/4, width/4, false);
        menu[2]=Bitmap.createScaledBitmap(b[8][0], width/4, width/4, false);
    }


    static class MapTitle{
        private final Pair<Integer, Integer> coords;
        private Bitmap texture;
        private int type;

        public MapTitle(Pair<Integer, Integer> coords, Bitmap texture, int type) {
            this.coords = coords;
            this.texture = texture;
            this.type = type;
        }

        public Pair<Integer, Integer> getCoords() {
            return coords;
        }

        public void setTexture(Bitmap texture) {
            this.texture = Bitmap.createBitmap(texture);
        }

        public Bitmap getTexture() {
            return texture;
        }

        public int getType() {
            return type;
        }

        public void setType(int type) {
            this.type = type;
        }
    }
}

File: FinalProject/app/src/main/java/com/example/finalproject/MainActivity.kt
package com.example.finalproject

import android.content.Context
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.graphics.Point
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.WindowInsets
import android.view.WindowManager
import androidx.appcompat.app.AppCompatActivity
import androidx.room.ColumnInfo
import androidx.room.Dao
import androidx.room.Database
import androidx.room.Entity
import androidx.room.Insert
import androidx.room.PrimaryKey
import androidx.room.Query
import androidx.room.Room
import androidx.room.RoomDatabase
import com.example.finalproject.fragments.MapFragment
import com.example.finalproject.fragments.MenuFragment
import com.example.finalproject.fragments.StatusBarFragment
import com.example.finalproject.fragments.TutorialFragment
import com.example.finalproject.service.*
import com.example.finalproject.service.classes.*
import com.example.finalproject.service.classes.Map
import com.example.finalproject.service.classes.entities.Enemy
import com.example.finalproject.service.classes.entities.Player
import com.example.finalproject.service.classes.items.Item
import com.example.finalproject.service.classes.spell.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import org.xmlpull.v1.XmlPullParser
import kotlin.concurrent.thread

class MainActivity : AppCompatActivity() {

    /**
     * initializes base data required for app to function, calls graphic components
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        goFullscreen()
        val fragmentTransaction = supportFragmentManager.beginTransaction()
        var mapParser: XmlPullParser = resources.getXml(R.xml.start_map)
        map.add(Map(mapParser))
        mapParser = resources.getXml(R.xml.first_village_map)
        map.add(Map(mapParser))
        val bounds = getBounds()
        val sharedPreference = getSharedPreferences("PREFERENCE_NAME", Context.MODE_PRIVATE)
        saved = sharedPreference.getBoolean("saved", false)
        width = bounds.first
        height = bounds.second
        res = resources
        music = Music()
        music.start(this, R.raw.main)
        player = Player(2, 2)
        db = Room.databaseBuilder(
            applicationContext,
            AppDatabase::class.java,
            "saves"
        ).build()
        setTextures()
        setInitialData(saved)
        while (!inited)
            continue
        if (!saved) {
            fragmentTransaction.add(R.id.tutorial, TutorialFragment())
            showTutorial = false
        } else {
            fragmentTransaction.add(R.id.map, MapFragment(player.mapNumber))
            fragmentTransaction.add(R.id.status, StatusBarFragment())
            fragmentTransaction.add(R.id.menu, MenuFragment())
        }
        fragmentTransaction.commit()
    }

    override fun onStart() {
        super.onStart()
        music.start(this, R.raw.main)
        val sharedPreference = getSharedPreferences("PREFERENCE_NAME", Context.MODE_PRIVATE)
        saved = sharedPreference.getBoolean("saved", false)
    }

    override fun onStop() {
        super.onStop()
        music.stop()
        val sharedPreference = getSharedPreferences("PREFERENCE_NAME", Context.MODE_PRIVATE)
        val editor = sharedPreference.edit()
        editor.putBoolean("saved", saved)
        editor.apply()
    }

    @Entity
    class Saves(
        @PrimaryKey val id: Int,
        @ColumnInfo(name = "assets") val assetsSave: String,
        @ColumnInfo(name = "player") val playerSave: String,
        @ColumnInfo(name = "tutorial") val tutorialSave: Boolean
    )

    @Dao
    interface SavesDao {

        /**
         * @return saves from database
         */
        @Query("SELECT * FROM saves")
        fun getJsons(): List<Saves>

        /**
         * clears the save
         */
        @Query("DELETE FROM saves WHERE id=0")
        fun deleteSaves()

        /**
         * @param save inserts the save into the database
         */
        @Insert
        fun insertSave(save: Saves)
    }

    @Database(entities = [Saves::class], version = 1)
    abstract class AppDatabase : RoomDatabase() {
        abstract fun userDao(): SavesDao
    }

    companion object {
        lateinit var player: Player
        private lateinit var res: Resources
        private var width: Int = 0
        private var height: Int = 0
        private var menuWidth = 0
        private var mapTitleWidth = 0
        var avatarWidth = 0
        var statusImagesWidth = 0
        var categoryImageWidth = 0
        val map = ArrayList<Map>()
        val menu = arrayOfNulls<Bitmap>(4)
        private val mapTextures = ArrayList<Bitmap>()
        private lateinit var avatar: Bitmap
        lateinit var textures: Array<Array<Bitmap>>
        lateinit var music: Music
        lateinit var assets: Assets
        var showTutorial = true
        var saved = false
        lateinit var db: AppDatabase
        private var inited = false

        @Serializable
        class Assets {
            val chancesOfFight =
                HashMap<Int, Int>() //<id of a location, chance of getting into a fight>
            val chancesOfEnemy =
                HashMap<Int, ArrayList<Pair<Int, Int>>>() //<id of a location, <chance of getting into a fight with that enemy, id of an enemy>
            val enemies = HashMap<Int, Enemy>() //<id of an enemy, object template>
            val components = HashMap<Int, Component>() //<id of a component, object template>
            val elements = ArrayList<Int>()
            val manaChannels = ArrayList<Int>()
            val types = ArrayList<Int>()
            val forms = ArrayList<Int>()
            val manaReservoirs = ArrayList<Int>()
            val researches = HashMap<Int, Research>() //<id of a research, object template>
            val researchEffects =
                HashMap<Int, ResearchEffect>() //<id of an effect, object template>
            val availableResearches = ArrayList<Int>() //<id of a research>
            val recipes = ArrayList<Recipe>()
            val items = HashMap<Int, Item>() //<id of an item, object template>
            val shopList = ArrayList<Item>()
            val tasks = HashMap<Int, Task>() //<id of a task, object template>
            val activeTasks = ArrayList<Int>()
            val itemsObtained: HashMap<Int, Int> = HashMap() //<id of an item, quantity>
            val enemiesKilled: HashMap<Int, Int> = HashMap() //<id of an enemy, quantity>
        }

        /**
         * @return player's avatar
         */
        fun getAvatar(): Bitmap = Bitmap.createBitmap(avatar)

        /**
         * parses textures and initializes map
         */
        private fun setTextures() {
            val texturesSource: Bitmap
            val xy = width * 1.0 / height
            if (xy in 0.4..0.6) {
                menuWidth = width / 4
                avatarWidth = width / 3
                mapTitleWidth = width * 1000 / 5000
            } else {
                mapTitleWidth = height * 1000 / 9000
                menuWidth = width / 4
                avatarWidth = width / 4
            }
            statusImagesWidth = width / 10
            categoryImageWidth = width / 4
            val n = 10
            val m = 20
            texturesSource = Bitmap.createScaledBitmap(
                BitmapFactory.decodeResource(res, R.drawable.textures),
                mapTitleWidth * m, mapTitleWidth * n, false
            )
            textures = Array(n) {
                Array(m) {
                    Bitmap.createBitmap(1, 1, Bitmap.Config.ARGB_8888)
                }
            }
            for (i in 0 until n) {
                for (j in 0 until m) {
                    textures[i][j] = Bitmap.createBitmap(
                        texturesSource,
                        j * mapTitleWidth,
                        i * mapTitleWidth,
                        mapTitleWidth,
                        mapTitleWidth
                    )
                }
            }
            mapTextures.addAll(textures[1].asList())
            val id = 512
            for (i in menu.indices)
                menu[i] = Bitmap.createScaledBitmap(
                    textures[0][i], menuWidth, menuWidth, false
                )
            for (i in map.indices)
                for (j in 0 until map[i].length)
                    for (k in 0 until map[i].width)
                        if (map[i].map[j][k].id != id + 255)
                            map[i].map[j][k].setTexture(
                                mapTextures[map[i].map[j][k].id - 512]
                            )
                        else {
                            map[i].map[j][k].setTexture(
                                Bitmap.createBitmap(
                                    mapTitleWidth,
                                    mapTitleWidth,
                                    Bitmap.Config.ARGB_8888
                                )
                            )
                            map[i].map[j][k].getTexture().eraseColor(Color.BLACK)
                        }
            avatar = Bitmap.createScaledBitmap(textures[5][5], mapTitleWidth, mapTitleWidth, false)
        }

        /**
         * @param saved true if the player has already made saves, false otherwise
         * initializes assets either from Room or from xml file
         */
        fun setInitialData(saved: Boolean = false) {
            if (saved) thread {
                val dao = db.userDao()
                val saves = dao.getJsons()
                assets =
                    Json.decodeFromString(Assets.serializer(), saves[saves.size - 1].assetsSave)
                player =
                    Json.decodeFromString(Player.serializer(), saves[saves.size - 1].playerSave)
                showTutorial = saves[saves.size - 1].tutorialSave
                inited = true
            } else {
                var data = ""
                val parser = res.getXml(R.xml.initial_data)
                while (parser.eventType != XmlPullParser.END_TAG) {
                    if (parser.eventType == XmlPullParser.START_TAG && parser.name == "data") {
                        parser.next()
                        data = parser.text
                    }
                    parser.next()
                }
                assets = Json.decodeFromString(Assets.serializer(), data)
                inited = true
                Log.d("Assets", Json.encodeToString(Assets.serializer(), assets))
            }
        }
    }

    /**
     * regulates app's window size
     */
    private fun goFullscreen() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            window.insetsController?.hide(WindowInsets.Type.statusBars())
        } else {
            window.setFlags(
                WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN
            )
        }
    }

    /**
     * @return width and height of a device's screen
     * calculates returned value to scale textures
     */
    @Suppress("DEPRECATION")
    private fun getBounds(): Pair<Int, Int> {
        val wm = windowManager
        val width: Int
        val height: Int
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            val windowMetrics = wm.currentWindowMetrics
            val windowInsets: WindowInsets = windowMetrics.windowInsets
            val insets = windowInsets.getInsetsIgnoringVisibility(
                WindowInsets.Type.navigationBars() or WindowInsets.Type.displayCutout()
            )
            val insetsWidth = insets.right + insets.left
            val insetsHeight = insets.top + insets.bottom

            val bounds = windowMetrics.bounds
            width = bounds.width() - insetsWidth
            height = bounds.height() - insetsHeight
        } else {
            val size = Point()
            val display = wm.defaultDisplay
            display?.getSize(size)
            width = size.x
            height = size.y
        }
        return Pair(width, height);
    }
}